python/lsst/ts/scheduler/scheduler_csc.py

```diff
@@ -493,3 +715,177 @@ def run(self):
+                retval = False
+            else:
+                self.log.error('Unrecognized state [%i] for observation %i for target %s.',
+                               info.processState,
+                               check[1],
+                               check[0])
+                retval = False
+
+        return retval
+
+    async def simple_target_production_loop(self):
+        """ This coroutine implements the simple target production loop. It will query the status of the queue and,
+        if there is nothing running, it will add an observation to the back of the queue. Once Scheduler is enabled
+        with simple mode, this coroutine will be added to the event loop as a future. While running it will checking
+        the queue and put targets to it. If the queue is paused or there is a running script, it will not add any
+        target to it. If the Scheduler is disabled this method will be cancelled.
+
+        Also, note that the target production loop does not care about if it is day or night. It is up to the
+        scheduling algorithm to decide whether there is something to observe or not given the input telemetry. A
+        target produced by the scheduler can have a start time. In this case, the task will be sent to the queue
+        and it is up to the queue or the running script to wait for the specified time. If the scheduling algorithm
+        does not produce any target, then the Scheduler will keep updating telemetry and requesting targets until
+        something is returned to be executed.
+
+        NOTE: This method may either run in the main event loop or in an event loop of a thread. If it is the first
+        case, we need to make sure long running tasks are sent to a thread or a process. If the latter, then it
+        should be ok to simple execute long running tasks and block the thread event loop.
+
+        """
+
+        self.run_loop = True
+        self.raw_telemetry['scheduled_targets'] = []
+
+        loop = asyncio.get_event_loop()
+
+        first_pass = True
+        while self.summary_state == base_csc.State.ENABLED and self.run_loop:
+            try:
+                # Only request queue status on the first pass. After that, this call will block until the
+                # queue status changes.
+                queue = await self.get_queue(request=first_pass)
+
+                # This return False if script failed, in which case get_queue will request status from the queue.
+                first_pass = not await self.check_scheduled()
+
+                # Only send targets to queue if it is running, empty and there is nothing executing
+                if queue.running and queue.currentSalIndex == 0 and queue.length == 0:
+                    # TODO: publish detailed state indicating that the scheduler is selecting a target
+
+                    self.log.log(WORDY, 'Queue ready to receive targets.')
+
+                    await loop.run_in_executor(None, self.update_telemetry)
```

Comment:
So, the scheduler will actually not get the telemetry data from the SAL but rather construct a telemetry stream by querying the EFD. We do that because we need more than just the latest value and it seems safer to query the EFD instead of having the scheduler store that data itself.